subprojects/core/src/integTest/groovy/org/gradle/api/tasks/FailingIncrementalTasksIntegrationTest.groovy

```diff
@@ -33,15 +34,83 @@ class FailingIncrementalTasksIntegrationTest extends AbstractIntegrationSpec {
             }
         """
 
+        expect:
+        succeeds "foo"
+
         when:
-        run "foo"
         file("out.txt") << "force rerun"
-        def failure1 = runAndFail "foo"
-        def failure2 = runAndFail "foo"
+        fails"foo"
+        then:
+        failureHasCause "Boo!"
 
+        when:
+        fails "foo", "--info"
         then:
-        failure1.assertHasCause("Boo!")
-        failure2.assertHasCause("Boo!")
+        failureHasCause "Boo!"
+        output.contains "Task has failed previously."
         //this exposes an issue we used to have with in-memory cache.
     }
+
+    @Unroll
+    def "incremental task after previous #failiureCount failure(s) #description"() {
```

Comment:
The new thing here is that a build can fail in between and mess with the history. This could have happened before, too. I see no test testing failures between builds in `IncrementalTasksIntegrationTest`. We should probably add those. WDYT?